File: backend/src/ai/policy.ts
import type { Direction, ShipActionRequest } from "@shared/index.js";
import type { Game } from "../core/Game.js";
import type { Ship } from "../models/Ship.js";
import { rowCol } from "../core/coords.js";
import { Targeter, missCenter } from "./targeting.js";

/** Decide the next single action for a non-submarine ship. */
export function pickAttackAction(
  ship: Ship,
  targeter: Targeter,
): ShipActionRequest | null {
  // For aircraft carrier / cruiser / frigate (SINGLE_HIT only).
  if (ship.hasQuota("SINGLE_HIT")) {
    const target = targeter.bestSingleTarget();
    if (target < 0) return null;
    return { shipId: ship.id, kind: "SINGLE_HIT", targets: [target] };
  }
  // Battleship (AREA_HIT_2X2 only).
  if (ship.hasQuota("AREA_HIT_2X2")) {
    const { anchor } = targeter.bestAreaTarget();
    return { shipId: ship.id, kind: "AREA_HIT_2X2", targets: [anchor] };
  }
  return null;
}

/**
 * Pick a single action for the submarine: hit, move, or rotate. The submarine
 * has only one action per turn (shared budget); this function is called at
 * most once.
 */
export function pickSubmarineAction(
  game: Game,
  submarine: Ship,
  targeter: Targeter,
): ShipActionRequest | null {
  if (!submarine.hasQuota("SINGLE_HIT")) return null; // no budget

  // Decide between SINGLE_HIT vs evasive action.
  const enemyView = game.aiBoard.toEnemyView(); // what HUMAN sees of AI
  const subCells = submarine.positions;
  // Count how many cells adjacent to the submarine the HUMAN has attacked
  // (those tell us the human is "feeling around" us).
  let pressure = 0;
  for (const cell of subCells) {
    const { row, col } = rowCol(cell, enemyView.size);
    for (const [dr, dc] of [
      [-1, 0],
      [1, 0],
      [0, -1],
      [0, 1],
    ] as const) {
      const r = row + dr;
      const c = col + dc;
      if (r < 0 || r >= enemyView.size || c < 0 || c >= enemyView.size) continue;
      const k = enemyView.cells[r * enemyView.size + c];
      if (k === "MISS" || k === "HIT") pressure += 1;
    }
  }

  // If the submarine is itself damaged, it's already been spotted.
  const damaged = submarine.hits.size > 0;

  // Heuristic: under heavy pressure or damaged, try evasion.
  if ((pressure >= 3 || damaged) && tryEvade(game, submarine) != null) {
    const choice = tryEvade(game, submarine)!;
    return choice;
  }

  // Otherwise, attack.
  const target = targeter.bestSingleTarget();
  if (target < 0) return null;
  return { shipId: submarine.id, kind: "SINGLE_HIT", targets: [target] };
}

/**
 * Try to find a legal move/rotate for the submarine. Move is preferred over
 * rotate because moving disrupts the human's targeting more.
 */
function tryEvade(game: Game, submarine: Ship): ShipActionRequest | null {
  // Move away from the most-attacked side (use miss-center as proxy).
  const myView = game.aiBoard.toEnemyView();
  const center = missCenter(myView);
  const subAnchor = submarine.positions[0]!;
  const { row: sr, col: sc } = rowCol(subAnchor, myView.size);

  const candidates: Direction[] = [];
  if (center) {
    // Move opposite to the centroid direction.
    if (sr < center.row) candidates.push("N");
    else candidates.push("S");
    if (sc < center.col) candidates.push("W");
    else candidates.push("E");
  } else {
    // No data: try cardinal order N E S W.
    candidates.push("N", "E", "S", "W");
  }
  for (const d of candidates) {
    if (canMove(game, submarine, d)) {
      return { shipId: submarine.id, kind: "MOVE_1", targets: [], direction: d };
    }
  }
  if (canRotate(game, submarine)) {
    return { shipId: submarine.id, kind: "ROTATE_90", targets: [] };
  }
  return null;
}

function canMove(game: Game, submarine: Ship, dir: Direction): boolean {
  // Simulate by trying canPlace with translated cells.
  const size = game.aiBoard.size;
  const offset = (() => {
    switch (dir) {
      case "N":
        return -size;
      case "S":
        return size;
      case "W":
        return -1;
      case "E":
        return 1;
    }
  })();
  const newCells = submarine.positions.map((c) => c + offset);
  // Bounds: dir N/S keeps column constant; E/W keeps row constant.
  for (let i = 0; i < submarine.positions.length; i++) {
    const oldRC = rowCol(submarine.positions[i]!, size);
    const newRC = rowCol(newCells[i]!, size);
    if (dir === "E" || dir === "W") {
      if (oldRC.row !== newRC.row) return false; // wrapped
      if (newRC.col < 0 || newRC.col >= size) return false;
    }
    if (dir === "N" || dir === "S") {
      if (oldRC.col !== newRC.col) return false;
      if (newRC.row < 0 || newRC.row >= size) return false;
    }
  }
  if (!game.aiBoard.canPlaceCells(newCells, submarine)) return false;
  // Destination cells must currently be UNKNOWN/SHIP from the AI's perspective.
  for (const c of newCells) {
    const k = game.aiBoard.cells[c];
    if (k === "MISS" || k === "HIT" || k === "SUNK") return false;
  }
  return true;
}

function canRotate(game: Game, submarine: Ship): boolean {
  const anchor = submarine.positions[0]!;
  const size = game.aiBoard.size;
  const newOrient = submarine.orientation === "horizontal" ? "vertical" : "horizontal";
  const cells: number[] = [];
  for (let i = 0; i < submarine.length; i++) {
    const { row, col } = rowCol(anchor, size);
    const r = newOrient === "horizontal" ? row : row + i;
    const c = newOrient === "horizontal" ? col + i : col;
    if (r < 0 || r >= size || c < 0 || c >= size) return false;
    cells.push(r * size + c);
  }
  if (!game.aiBoard.canPlaceCells(cells, submarine)) return false;
  for (let i = 1; i < cells.length; i++) {
    const k = game.aiBoard.cells[cells[i]!];
    if (k === "MISS" || k === "HIT" || k === "SUNK") return false;
  }
  return true;
}

File: backend-rs/src/ai/policy.rs
use crate::core::coords::row_col;
use crate::core::game::Game;
use crate::models::Ship;
use crate::shared::{ActionKind, CellKnowledge, Direction, ShipActionRequest};

use super::targeting::{miss_center, Targeter};

/// Decide the next single action for a non-submarine ship.
pub fn pick_attack_action(ship: &Ship, targeter: &Targeter) -> Option<ShipActionRequest> {
    if ship.has_quota(ActionKind::SingleHit) {
        let target = targeter.best_single_target();
        if target < 0 {
            return None;
        }
        return Some(ShipActionRequest {
            ship_id: ship.id.clone(),
            kind: ActionKind::SingleHit,
            targets: vec![target as u32],
            direction: None,
        });
    }
    if ship.has_quota(ActionKind::AreaHit2x2) {
        let (anchor, _) = targeter.best_area_target();
        return Some(ShipActionRequest {
            ship_id: ship.id.clone(),
            kind: ActionKind::AreaHit2x2,
            targets: vec![anchor],
            direction: None,
        });
    }
    None
}

/// Pick a single action for the submarine: hit, move, or rotate.
pub fn pick_submarine_action(
    game: &Game,
    submarine: &Ship,
    targeter: &Targeter,
) -> Option<ShipActionRequest> {
    if !submarine.has_quota(ActionKind::SingleHit) {
        return None;
    }

    let enemy_view = game.ai_board.to_enemy_view(); // what HUMAN sees of AI
    let sub_cells = &submarine.positions;
    let mut pressure = 0;
    for cell in sub_cells {
        let (row, col) = row_col(*cell, enemy_view.size);
        for (dr, dc) in [(-1, 0), (1, 0), (0, -1), (0, 1)] {
            let r = row + dr;
            let c = col + dc;
            if r < 0 || r >= enemy_view.size as i32 || c < 0 || c >= enemy_view.size as i32 {
                continue;
            }
            let k = enemy_view.cells[(r as u32 * enemy_view.size + c as u32) as usize];
            if matches!(k, CellKnowledge::Miss | CellKnowledge::Hit) {
                pressure += 1;
            }
        }
    }

    let damaged = !submarine.hits.is_empty();

    if (pressure >= 3 || damaged) && try_evade(game, submarine).is_some() {
        return try_evade(game, submarine);
    }

    let target = targeter.best_single_target();
    if target < 0 {
        return None;
    }
    Some(ShipActionRequest {
        ship_id: submarine.id.clone(),
        kind: ActionKind::SingleHit,
        targets: vec![target as u32],
        direction: None,
    })
}

fn try_evade(game: &Game, submarine: &Ship) -> Option<ShipActionRequest> {
    let my_view = game.ai_board.to_enemy_view();
    let center = miss_center(&my_view);
    let sub_anchor = submarine.positions[0];
    let (sr, sc) = row_col(sub_anchor, my_view.size);

    let mut candidates: Vec<Direction> = Vec::new();
    if let Some((cr, cc)) = center {
        if (sr as f64) < cr {
            candidates.push(Direction::N);
        } else {
            candidates.push(Direction::S);
        }
        if (sc as f64) < cc {
            candidates.push(Direction::W);
        } else {
            candidates.push(Direction::E);
        }
    } else {
        candidates.extend([Direction::N, Direction::E, Direction::S, Direction::W]);
    }
    for d in candidates {
        if can_move(game, submarine, d) {
            return Some(ShipActionRequest {
                ship_id: submarine.id.clone(),
                kind: ActionKind::Move1,
                targets: vec![],
                direction: Some(d),
            });
        }
    }
    if can_rotate(game, submarine) {
        return Some(ShipActionRequest {
            ship_id: submarine.id.clone(),
            kind: ActionKind::Rotate90,
            targets: vec![],
            direction: None,
        });
    }
    None
}

fn can_move(game: &Game, submarine: &Ship, dir: Direction) -> bool {
    let size = game.ai_board.size;
    let offset: i32 = match dir {
        Direction::N => -(size as i32),
        Direction::S => size as i32,
        Direction::W => -1,
        Direction::E => 1,
    };
    let new_cells: Vec<i32> = submarine
        .positions
        .iter()
        .map(|c| (*c as i32) + offset)
        .collect();

    for (i, _) in submarine.positions.iter().enumerate() {
        if new_cells[i] < 0 || (new_cells[i] as u32) >= size * size {
            return false;
        }
        let (old_r, old_c) = row_col(submarine.positions[i], size);
        let (new_r, new_c) = row_col(new_cells[i] as u32, size);
        if matches!(dir, Direction::E | Direction::W) {
            if old_r != new_r {
                return false;
            }
            if new_c < 0 || new_c >= size as i32 {
                return false;
            }
        }
        if matches!(dir, Direction::N | Direction::S) {
            if old_c != new_c {
                return false;
            }
            if new_r < 0 || new_r >= size as i32 {
                return false;
            }
        }
    }
    let sub_idx = game
        .ai_board
        .ships
        .iter()
        .position(|s| s.id == submarine.id);
    let new_cells_u32: Vec<u32> = new_cells.iter().map(|c| *c as u32).collect();
    if !game.ai_board.can_place_cells(&new_cells_u32, sub_idx) {
        return false;
    }
    for c in &new_cells_u32 {
        let k = game.ai_board.cells[*c as usize];
        if matches!(
            k,
            CellKnowledge::Miss | CellKnowledge::Hit | CellKnowledge::Sunk
        ) {
            return false;
        }
    }
    true
}

fn can_rotate(game: &Game, submarine: &Ship) -> bool {
    let anchor = submarine.positions[0];
    let size = game.ai_board.size;
    let new_orient = match submarine.orientation {
        crate::shared::Orientation::Horizontal => crate::shared::Orientation::Vertical,
        crate::shared::Orientation::Vertical => crate::shared::Orientation::Horizontal,
    };
    let mut cells: Vec<u32> = Vec::new();
    for i in 0..(submarine.length() as i32) {
        let (row, col) = row_col(anchor, size);
        let (r, c) = match new_orient {
            crate::shared::Orientation::Horizontal => (row, col + i),
            crate::shared::Orientation::Vertical => (row + i, col),
        };
        if r < 0 || r >= size as i32 || c < 0 || c >= size as i32 {
            return false;
        }
        cells.push((r as u32) * size + (c as u32));
    }
    let sub_idx = game
        .ai_board
        .ships
        .iter()
        .position(|s| s.id == submarine.id);
    if !game.ai_board.can_place_cells(&cells, sub_idx) {
        return false;
    }
    for i in 1..cells.len() {
        let k = game.ai_board.cells[cells[i] as usize];
        if matches!(
            k,
            CellKnowledge::Miss | CellKnowledge::Hit | CellKnowledge::Sunk
        ) {
            return false;
        }
    }
    true
}
